sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,553 @@
+                SELECT
+                    C.column_name, C.ordinal_position, C.data_type,
+                    C.character_maximum_length, C.numeric_precision, C.numeric_scale,
+                    KCU.ordinal_position AS pk_ordinal_position
+                FROM information_schema.columns C
+                LEFT OUTER JOIN information_schema.key_column_usage KCU
+                ON (KCU.table_schema = C.table_schema)
+                AND (KCU.table_name = C.table_name)
+                AND (KCU.column_name = C.column_name)
+                WHERE C.table_schema=:table_schema AND C.table_name=:table_name
+                ORDER BY C.ordinal_position;
+            """
+            r = sess.execute(
+                table_info_sql,
+                {"table_schema": self.db_schema, "table_name": dataset.table_name},
+            )
+            ms_table_info = list(r)
+
+            id_salt = f"{self.db_schema} {dataset.table_name} {self.get_db_tree()}"
+            schema = sqlserver_adapter.sqlserver_to_v2_schema(ms_table_info, id_salt)
+            yield "schema.json", schema.to_column_dicts()
+
+    _UNSUPPORTED_META_ITEMS = (
+        "title",
+        "description",
+        "metadata/dataset.json",
+        "metadata.xml",
+    )
+
+    @classmethod
+    def try_align_schema_col(cls, old_col_dict, new_col_dict):
+        old_type = old_col_dict["dataType"]
+        new_type = new_col_dict["dataType"]
+
+        # Some types have to be approximated as other types in SQL Server.
+        if sqlserver_adapter.APPROXIMATED_TYPES.get(old_type) == new_type:
+            new_col_dict["dataType"] = new_type = old_type
+
+        # Geometry type loses its extra type info when roundtripped through SQL Server.
+        if new_type == "geometry":
+            new_col_dict["geometryType"] = old_col_dict.get("geometryType")
+            new_col_dict["geometryCRS"] = old_col_dict.get("geometryCRS")
+
+        return new_type == old_type
+
+    def _remove_hidden_meta_diffs(self, dataset, ds_meta_items, wc_meta_items):
+        super()._remove_hidden_meta_diffs(dataset, ds_meta_items, wc_meta_items)
+
+        # Nowhere to put these in postgis WC
+        for key in self._UNSUPPORTED_META_ITEMS:
+            if key in ds_meta_items:
+                del ds_meta_items[key]
+
+        # Diffing CRS is not yet supported.
+        for key in list(ds_meta_items.keys()):
+            if key.startswith("crs/"):
+                del ds_meta_items[key]
+
+    def _is_meta_update_supported(self, dataset_version, meta_diff):
+        """
+        Returns True if the given meta-diff is supported *without* dropping and rewriting the table.
+        (Any meta change is supported if we drop and rewrite the table, but of course it is less efficient).
+        meta_diff - DeltaDiff object containing the meta changes.
+        """
+        # For now, just always drop and rewrite.
+        return not meta_diff
+
+
+class InstanceFunction(Function):
```

Comment:
This is generic to MSSQL but not our working copy? Would it be better homed in `sqlserver_adapter.py`?

Reply:
Addressed below (for some reason github collapsed your upper 12 comments, which is why I missed them and addressed the lower ones first).

Seems happy where it is for right now but happy to move more SQL helpers out of working-copies as and when we need them for importers or testing.